pilot/pkg/model/discoverability.go

```diff
@@ -0,0 +1,34 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package model
+
+// ServiceDiscoverabilityPolicy manages discoverability of service endpoints across the mesh.
+type ServiceDiscoverabilityPolicy interface {
+	// IsDiscoverableFromProxy indicates whether the given service endpoint should be
+	// made discoverable to the given proxy.
+	IsDiscoverableFromProxy(*Service, *IstioEndpoint, *Proxy) bool
```

Comment:
I'm a bit worried about the model - and exposing those interfaces ( which are really specific to a particular implementation ). I assume this API is needed to support 'primary' mode, where we hold all endpoints from the mesh in memory and then filter based on the caller. 

In multi-primary - this is too inefficient, instead we can just drop all un-exported  proxies at load time.

We could also store exported endpoints in a different index, which is also efficient. 

I am ok if we don't optimize now - but I'm concerned that defining this as an interface in model will make it harder to change. Can we at least keep this interface next to the implementation ( since 'model' package should really be used for common interfaces ), and make it clear that is only used with this particular filtering impl ? 

Unless I missed how it works - in which case some comments would clarify ...

Reply:
I'm not sure I see how this interface implies something about the implementation. This same interface could be used for our cluster-local implementation (which is stored in MeshConfig today).

Can you expand a bit on the inefficiencies that this API introduces?

Reply:
This comment no longer applies.